Dart/src/com/jetbrains/lang/dart/ide/template/DartTemplateContextType.java

```diff
@@ -1,27 +1,62 @@
 package com.jetbrains.lang.dart.ide.template;
 
+import com.intellij.codeInsight.template.EverywhereContextType;
 import com.intellij.codeInsight.template.TemplateContextType;
 import com.intellij.openapi.fileTypes.SyntaxHighlighter;
+import com.intellij.psi.PsiElement;
 import com.intellij.psi.PsiFile;
+import com.intellij.psi.util.PsiTreeUtil;
 import com.jetbrains.lang.dart.DartBundle;
 import com.jetbrains.lang.dart.DartLanguage;
 import com.jetbrains.lang.dart.highlight.DartSyntaxHighlighter;
+import com.jetbrains.lang.dart.psi.DartBlock;
+import org.jetbrains.annotations.NonNls;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-public class DartTemplateContextType extends TemplateContextType {
-  protected DartTemplateContextType() {
-    super("DART", DartBundle.message("dart.language.id"));
+public abstract class DartTemplateContextType extends TemplateContextType {
+  protected DartTemplateContextType(@NotNull @NonNls String id,
+                                    @NotNull String presentableName,
+                                    @Nullable Class<? extends TemplateContextType> baseContextType) {
+    super(id, presentableName, baseContextType);
   }
 
   @Override
   public boolean isInContext(@NotNull PsiFile file, int offset) {
-    return file.getLanguage() instanceof DartLanguage;
+    if (file.getLanguage() instanceof DartLanguage) {
+      PsiElement element = file.findElementAt(offset);
+      return element != null && isInContext(element);
+    }
+    return false;
   }
 
+  protected abstract boolean isInContext(@NotNull PsiElement element);
+
   @Nullable
   @Override
   public SyntaxHighlighter createHighlighter() {
     return new DartSyntaxHighlighter();
   }
+
+  public static class Generic extends DartTemplateContextType {
+    public Generic() {
+      super("DART_CODE", DartBundle.message("dart.language.id"), EverywhereContextType.class);
```

Comment:
User defined live templates configured to be applicable in Dart will stop working if we change 'Dart' serialization to 'DART_CODE'. Are there downsides of keeping 'DART'?

See also `dart_surround.xml` file: templates defined there do not work with this pull request.
BTW `dart_surround.xml` itself looks suspicious as it introduces a separate templates group. Not sure if adding more than one group per language is justified. Also I never used templates from this group, may be this is a dead horse. We can look for similar things in other languages for reference.